core/src/main/java/org/mule/runtime/core/internal/transformer/graph/TransformationVertex.java

```diff
@@ -0,0 +1,48 @@
+/*
+ * Copyright (c) MuleSoft, Inc.  All rights reserved.  http://www.mulesoft.com
+ * The software in this package is published under the terms of the CPAL v1.0
+ * license, a copy of which has been included with this distribution in the
+ * LICENSE.txt file.
+ */
+package org.mule.runtime.core.internal.transformer.graph;
+
+import org.mule.runtime.api.metadata.DataType;
+
+
+/**
+ * Helper POJO to represent vertex in a {@link TransformationGraph}. It handles special behaviour when comparing a hashing different {@link DataType}s so that
+ * {@link DataType}s with compatible {@link org.mule.runtime.api.metadata.MediaType}s are considered equals.
+ */
+public class TransformationVertex {
+
+  private DataType dataType;
+
+  public TransformationVertex(DataType dataType) {
+    this.dataType = dataType;
+  }
+
+  public DataType getDataType() {
+    return dataType;
+  }
+
+  public void setDataType(DataType dataType) {
+    this.dataType = dataType;
+  }
+
+  @Override
+  public boolean equals(Object obj) {
+    if (this == obj) {
+      return true;
+    }
+    if (obj.getClass() != TransformationVertex.class) {
+      return false;
+    }
+    return DataType.match(this.dataType, ((TransformationVertex) obj).dataType);
```

Comment:
this just moves the problem from `DataType` into here. A structure that doesn't rely on hash/equals must be used